src/hotspot/share/prims/scopedMemoryAccess.cpp

```diff
@@ -60,36 +77,42 @@ static bool for_scoped_method(JavaThread* jt, const Func& func) {
 #endif
 
     if (is_scoped) {
-      assert(depth < max_critical_stack_depth, "can't have more than %d critical frames", max_critical_stack_depth);
-      return func(stream);
-    }
-    depth++;
-
-#ifndef ASSERT
-    // On debug builds, just keep searching the stack
-    // in case we missed an @Scoped method further up
-    if (depth >= max_critical_stack_depth) {
-      break;
+      assert(!would_have_bailed, "would have missed scoped method on release build");
+      bool done = func(stream);
+      if (done || !agents_loaded) {
+        // We may also have to keep walking after finding a @Scoped method,
+        // since there may be multiple @Scoped methods active on the stack
+        // if a JVMTI agent callback runs during a scoped access and calls
+        // back into Java code that then itself does a scoped access.
+        return;
+      }
     }
-#endif
   }
-  return false;
 }
 
 static bool is_accessing_session(JavaThread* jt, oop session, bool& in_scoped) {
-  return for_scoped_method(jt, [&](vframeStream& stream){
+  if (jt->is_throwing_unsafe_access_error()) {
```

Comment:
If there are JVMTI agents, this might return true but the target could be in a different nested scoped access. I think we should check we are in the no agent case before bailing out.

Reply:
My assumption is that once an unsafe access error is thrown, we don't expect execution to continue. I suppose it is technically possible to catch the exception, either in Java code or in the native agent code, and then try to continue execution, but the program would be in an undefined state at that point already. In other words, I don't think anyone should be continuing execution after this exception happens.

Although, it doesn't seem like a bad idea to keep walking here as a fail safe.

Reply:
> My assumption is that once an unsafe access error is thrown, we don't expect execution to continue. I suppose it is technically possible to catch the exception, either in Java code or in the native agent code, and then try to continue execution, but the program would be in an undefined state at that point already. In other words, I don't think anyone should be continuing execution after this exception happens.
>
Yes, makes sense.